### PR TITLE
[GSIP-149] Improved GeoServer Style Page

### PR DESCRIPTION
**Pull request by tbarsballe**
See https://github.com/geoserver/geoserver/wiki/GSIP-149

With FOSS4G Bonn coming up, I want to try and get this into a 2.10 milestone release prior to Bonn.

Due to some time constraints and unforeseen complications, there are a few 'nice to have' features that didn't make it into this PR.
* The map preview still uses OL2 (not OL3)
* There is no link to preview as SLD (This functionality is covered by the REST API)
* There is no link to the styling docs (We can still search the docs manually)

Everything else from the proposal is included.

### CODE REVIEW

**Comment by jodygarnett** [comment]
Ran into an issue testing, added the note to [GEOS-7674|https://osgeo-org.atlassian.net/browse/GEOS-7674]

**Comment by tbarsballe** [comment]
Pushed up a fix for your issue @jodygarnett , I am not sure why I wasn't having the same problem.

(I am going to add incremental fixes as seperate commits for ease of use, and squash before merging)

File: src/web/wms/src/main/java/org/geoserver/wms/web/data/StyleAdminPanel.html [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+              <span><wicket:message key="legendSpan">Legend</wicket:message></span>
+            </legend>
+            <ul>
+              <li>
+                <label for="legend"><wicket:message key="legend">Legend</wicket:message></label>
+                <span wicket:id="legendPanel"></span>
+              </li>
+              <li>
+                <a href="#" wicket:id="preview"><wicket:message key="preview">Preview Legend</wicket:message></a>
+                <div wicket:id="legendContainer">
```
Comment (jodygarnett) [inline src/web/wms/src/main/java/org/geoserver/wms/web/data/StyleAdminPanel.html]: ⚠️ Legend container is modified a little bit by https://github.com/geoserver/geoserver/pull/1749

File: src/web/wms/src/main/java/org/geoserver/wms/web/data/BandsPanel.java [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+import org.geoserver.catalog.CoverageInfo;
+import org.geoserver.web.wicket.GeoServerDataProvider;
+import org.geoserver.web.wicket.GeoServerDataProvider.Property;
+import org.geoserver.web.wicket.GeoServerTablePanel;
+import org.geoserver.web.wicket.ParamResourceModel;
+import org.geotools.util.logging.Logging;
+import org.opengis.coverage.SampleDimensionType;
+
+@SuppressWarnings("serial")
+public class BandsPanel extends Panel {
```
Comment (jodygarnett) [inline src/web/wms/src/main/java/org/geoserver/wms/web/data/BandsPanel.java]: public class, want even a small javadoc comment :)

File: src/web/wms/src/main/java/org/geoserver/wms/web/data/DataAttribute.java [hunk 3]
```diff
@@ -0,0 +1,72 @@
+package org.geoserver.wms.web.data;
```
Comment (jodygarnett) [inline src/web/wms/src/main/java/org/geoserver/wms/web/data/DataAttribute.java]: header

File: src/web/wms/src/main/java/org/geoserver/wms/web/data/ExternalGraphicPanel.java [hunk 4]
```diff
@@ -292,6 +264,56 @@ protected String baseURL(Form form) {
...
+                
+                URI uri = new URI( external );
+                if( uri.isAbsolute() ){
+                    url = uri.toURL();
+                    if( !external.startsWith(baseUrl)){
+                        form.warn( "Recommend use of styles directory at "+baseUrl);
+                    }
+                }
+                else {
+                    url = new URL( baseUrl + "styles/"+external );
```
Comment (jodygarnett) [inline src/web/wms/src/main/java/org/geoserver/wms/web/data/ExternalGraphicPanel.java]: This logic changes a bit due to https://github.com/geoserver/geoserver/pull/1749

File: src/web/wms/src/main/java/org/geoserver/wms/web/data/LayerAssociationPanel.java [hunk 5]
```diff
@@ -0,0 +1,178 @@
...
+import org.apache.wicket.markup.html.panel.Fragment;
+import org.apache.wicket.markup.html.panel.Panel;
+import org.apache.wicket.model.IModel;
+import org.geoserver.catalog.LayerInfo;
+import org.geoserver.catalog.StyleInfo;
+import org.geoserver.web.wicket.GeoServerDataProvider;
+import static org.geoserver.web.wicket.GeoServerDataProvider.Property;
+import org.geoserver.web.wicket.GeoServerTablePanel;
+
+public class LayerAssociationPanel extends StyleEditTabPanel {
```
Comment (jodygarnett) [inline src/web/wms/src/main/java/org/geoserver/wms/web/data/LayerAssociationPanel.java]: public class, public javadoc.

I really like seeing this as a panel!

File: src/web/wms/src/main/java/org/geoserver/wms/web/data/LayerAttributePanel.java [hunk 6]
```diff
@@ -0,0 +1,69 @@
+package org.geoserver.wms.web.data;
```
Comment (jodygarnett) [inline src/web/wms/src/main/java/org/geoserver/wms/web/data/LayerAttributePanel.java]: header

File: src/web/wms/src/main/java/org/geoserver/wms/web/data/LayerAttributePanel.java [hunk 7]
```diff
@@ -0,0 +1,69 @@
...
+import org.apache.wicket.extensions.ajax.markup.html.modal.ModalWindow;
+import org.apache.wicket.model.Model;
+import org.apache.wicket.model.PropertyModel;
+import org.geoserver.catalog.CoverageInfo;
+import org.geoserver.catalog.FeatureTypeInfo;
+import org.geoserver.catalog.ResourceInfo;
+import org.geoserver.web.wicket.SimpleAjaxLink;
+import org.geotools.util.logging.Logging;
+
+public class LayerAttributePanel extends StyleEditTabPanel {
```
Comment (jodygarnett) [inline src/web/wms/src/main/java/org/geoserver/wms/web/data/LayerAttributePanel.java]: public classs, public javadoc

File: src/web/wms/src/main/java/org/geoserver/wms/web/data/StyleAdminPanel.java [hunk 8]
```diff
@@ -0,0 +1,498 @@
...
+                Session.get().error(new ParamResourceModel("styleNotFound", this, style.getFilename()).getString());
+            }
+        }
+        
+        // style generation functionality
+        templates = new DropDownChoice<StyleType>("templates", new Model<StyleType>(), new StyleTypeModel(), new StyleTypeChoiceRenderer());
+        templates.setOutputMarkupId(true);
+        templates.add(new AjaxFormComponentUpdatingBehavior("change") {
+
+            private static final long serialVersionUID = -6649152103570059645L;
```
Comment (jodygarnett) [inline src/web/wms/src/main/java/org/geoserver/wms/web/data/StyleAdminPanel.java]: Using small anonymous serializable commands is a bother with this java format, removes some of the terse

File: src/web/wms/src/main/java/org/geoserver/wms/web/data/StyleAdminPanel.java [hunk 9]
```diff
@@ -0,0 +1,498 @@
...
+     * Called when it is necessary to submit the form without needing to show validation, 
+     * such as when you are generating a new style
+     */
+    protected void clearFeedbackMessages() {
+        nameTextField.getFeedbackMessages().clear();
+        wsChoice.getFeedbackMessages().clear();
+        formatChoice.getFeedbackMessages().clear();
+        stylePage.editor.getFeedbackMessages().clear();
+    }
+    //TODO: Note - this does not reload the form...
```
Comment (jodygarnett) [inline src/web/wms/src/main/java/org/geoserver/wms/web/data/StyleAdminPanel.java]: Can we remove the TODO?

File: src/web/wms/src/main/resources/GeoServerApplication.properties [hunk 10]
```diff
@@ -1,23 +1,31 @@
...
-AbstractStylePage.styleFile        = Style file
-AbstractStylePage.upload           = Upload ...
-AbstractStylePage.validate         = Validate
-AbstractStylePage.preview          = Preview legend
-AbstractStylePage.styleNotFound    = Could not find Style file "{0}", please input a new style
-AbstractStylePage.genericError     = Invalid SLD
-AbstractStylePage.legend           = Legend
-ExtraStylesPalette.availableHeader = Available Styles
-ExtraStylesPalette.selectedHeader  = Selected Styles
+StyleAdminPanel.styleDataSpan       = Style Data
```
Comment (jodygarnett) [inline src/web/wms/src/main/resources/GeoServerApplication.properties]: Can we migrate the other language files? Are their any other language files ...

**Comment by jodygarnett** [comment]
Some additional feedbacks, mostly headers/javadocs. I am totally enjoying the functionality - and am glad we are going to get to play test this at foss4g.

File: src/web/wms/src/main/java/org/geoserver/wms/web/data/StyleAdminPanel.java (hunk 9 above)
Comment (tbarsballe) [inline src/web/wms/src/main/java/org/geoserver/wms/web/data/StyleAdminPanel.java]: Yes, I will do so. Got lost among a bunch of other things.

File: src/web/wms/src/main/resources/GeoServerApplication.properties (hunk 10 above)
Comment (tbarsballe) [inline src/web/wms/src/main/resources/GeoServerApplication.properties]: Yes there are other language files, but I think there is enough new content that we cannot readily migrate them with sufficient accuracy.

**Comment by tbarsballe** [comment]
Updated ExternalGraphicsPanel

**Comment by tbarsballe** [comment]
Manually squashed and merged to master.

**Comment by aaime** [comment]
Quite surprised to see this merged? GSIP approved gives the go for the work (it means it cannot be rejected anymore), not for merging the pull requests related to it. I told @jodygarnett that I was looking for time to review (I'm going to have time tomorrow). We recently have had GSIPs with pull requests that waited 20+ days and had no reviews, it seems we're looking at a fast land and a slow lane here...

**Comment by jodygarnett** [comment]
I reviewed this yesterday, prior to the GSIP being approved - indeed that was part of my feedback.

**Comment by tbarsballe** [comment]
(Also, going by the GSIP policies, this has been 'unoffically' approved for a few weeks, as we have had +1 from all but two of the PMC and no -1 since July 14, so technically should have been approved around July 21)

**Comment by aaime** [comment]
Guys, are you joking? All GSIPs past went though various steps, fist getting it voted, then start developing it, then get a review on the pull request. Unfortunately I made the mistake of telling @jodygarnett that I wanted to review only in private, yesterday, instead of stating it here, publicly (won't happen again). But seriously, look back and tell me which GSIP pull request got approved in 2 work days with just a quick review from a coworker (the weekend does not count, nobody is reviewing it anything in it anymore, since I stopped doing that). We always allowed some fair time for others to have a look, but it's not written anywhere, so I guess my next GSIP will be to make that explicit too.
